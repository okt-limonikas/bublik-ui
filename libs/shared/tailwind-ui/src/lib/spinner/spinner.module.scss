/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
@keyframes spinnerAnimation {
	0% {
		stroke-dashoffset: 33px;
		transform: rotatey(0deg) rotate(0deg);
	}
	24.9999% {
		stroke-dashoffset: 11px;
		transform: rotatey(0deg) rotate(0deg);
	}
	25% {
		stroke-dashoffset: 11px;
		transform: rotatey(180deg) rotate(270deg);
	}
	49.9999% {
		stroke-dashoffset: 33px;
		transform: rotatey(180deg) rotate(270deg);
	}
	50% {
		stroke-dashoffset: 33px;
		transform: rotatey(0deg) rotate(180deg);
	}
	74.9999% {
		stroke-dashoffset: 11px;
		transform: rotatey(0deg) rotate(180deg);
	}
	75% {
		stroke-dashoffset: 11px;
		transform: rotatey(180deg) rotate(90deg);
	}
	100% {
		stroke-dashoffset: 33px;
		transform: rotatey(180deg) rotate(90deg);
	}
}

@keyframes spinnerRotation {
	100% {
		transform: rotate(360deg);
	}
}

.spinner-wrapper {
	width: 100px;
	height: 100px;

	svg {
		width: 100%;
		height: 100%;
		overflow: visible;
		animation: spinnerRotation 1455ms infinite linear;

		circle {
			stroke: currentColor;
			stroke-dasharray: 44px;
			stroke-dashoffset: 11px;
			transform-origin: center;
			transform: rotatey(180deg) rotate(90deg);
			animation: spinnerAnimation 3850ms infinite ease;
		}
	}
}
