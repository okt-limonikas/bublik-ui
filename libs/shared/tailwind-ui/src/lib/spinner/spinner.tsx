/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { ComponentPropsWithRef, forwardRef } from 'react';

import { cn } from '../utils';

import styles from './spinner.module.scss';

export type SpinnerProps = ComponentPropsWithRef<'div'>;

export const Spinner = forwardRef<HTMLDivElement, SpinnerProps>(
	({ className, ...props }, ref) => {
		return (
			<div
				className={cn('grid place-items-center w-full text-primary', className)}
				{...props}
				ref={ref}
				data-testid="tw-spinner"
			>
				<div className={styles['spinner-wrapper']}>
					<svg
						width="16"
						height="16"
						viewBox="0 0 16 16"
						fill="none"
						xmlns="http://www.w3.org/2000/svg"
					>
						<circle cx="8" cy="8" r="7" strokeWidth="2" />
					</svg>
				</div>
			</div>
		);
	}
);
