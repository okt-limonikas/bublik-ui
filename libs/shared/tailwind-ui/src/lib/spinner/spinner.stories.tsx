/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { StoryFn, Meta } from '@storybook/react';
import { Spinner, SpinnerProps } from './spinner';

export default {
	component: Spinner,
	title: 'components/Spinner'
} as Meta;

export const Primary = {
	args: {}
};
