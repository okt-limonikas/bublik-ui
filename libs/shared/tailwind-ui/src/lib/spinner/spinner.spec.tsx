/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { it, describe, expect } from 'vitest';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { Spinner } from './spinner';

describe('Spinner', () => {
	it('should render successfully', () => {
		const { baseElement } = render(<Spinner />);
		expect(baseElement).toBeTruthy();
	});

	it('should match snapshot', () => {
		const tree = renderer.create(<Spinner />).toJSON();

		expect(tree).toMatchSnapshot();
	});
});
