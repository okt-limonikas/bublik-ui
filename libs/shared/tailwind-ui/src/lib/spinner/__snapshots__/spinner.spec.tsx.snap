// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Spinner > should match snapshot 1`] = `
<div
  className="grid place-items-center w-full text-primary"
  data-testid="tw-spinner"
>
  <div
    className="_spinner-wrapper_cbea93"
  >
    <svg
      fill="none"
      height="16"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="8"
        cy="8"
        r="7"
        strokeWidth="2"
      />
    </svg>
  </div>
</div>
`;
